parser.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ga4gh
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	oidc "github.com/coreos/go-oidc"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// Shim is used to convert an HTTP bearer authorization string that is _not_ in
+// the normal Identity format into an Identity.  This is useful when
+// interoperating with systems that do not yet provide a GA4GH identity.
+type Shim interface {
+	Shim(ctx context.Context, auth string) (*Identity, error)
+}
+
+// Parser parses authorization headers into Identity structs.  It first
+// attempts to use each of the Shims in order to convert the incoming identity.
+// If any of them succeed then the converted identity is returned immediately.
+// If no Shim converted the identity then Issuers is consulted.  Issuers is a
+// map of issuer URL to OAuth 2.0 Client ID.  If any of the issuers match the
+// issuer in the provided identity, then the token is validated against the
+// OIDC configuration that that issuer exposes.
+type Parser struct {
+	Shims   []Shim
+	Issuers map[string]string
+
+	mu      sync.Mutex
+	issuers map[string]*oidc.IDTokenVerifier
+}
+
+// Parse takes an authorization string (usually an HTTP authorization bearer
+// token) and converts it into an Identity.
+func (p *Parser) Parse(ctx context.Context, auth string) (*Identity, error) {
+	if err := p.lazyInit(ctx); err != nil {
+		return nil, fmt.Errorf("initializing parser: %v", err)
+	}
+
+	for _, shim := range p.Shims {
+		id, err := shim.Shim(ctx, auth)
+		if err == nil {
+			return id, nil
+		}
+	}
+
+	parsed, err := jwt.ParseSigned(auth)
+	if err != nil {
+		return nil, fmt.Errorf("parsing JWT: %v", err)
+	}
+
+	// This might look strange.  We're extracting the claims from the provided
```

Comment:
This comment isn't very helpful to me.  It does look strange, but this doesn't tell me *why* you are ignoring the signature or expiry values.  Aren't they still meaningful even if you don't yet know what claim you want to use?